### PR TITLE
HHH-13456 ForeignGenerator Throws ClassCastException When Using StatelessSession

### PR DESCRIPTION
**Pull request by NathanQingyangXu**
https://hibernate.atlassian.net/browse/HHH-13456

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/id/ForeignGenerator.java [hunk 1]
```diff
@@ -6,11 +6,12 @@
  */
 package org.hibernate.id;
 
+import static org.hibernate.internal.CoreLogging.messageLogger;
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/id/ForeignGenerator.java]: could you keep the static imports last please?

File: hibernate-core/src/main/java/org/hibernate/id/ForeignGenerator.java [hunk 2]
```diff
@@ -116,10 +115,12 @@ public Serializable generate(SharedSessionContractImplementor sessionImplementor
 			if (sessionImplementor instanceof Session) {
 				id = ((Session) sessionImplementor)
 						.save(foreignValueSourceType.getAssociatedEntityName(), associatedObject);
-			} else if (sessionImplementor instanceof StatelessSession) {
+			}
+			else if (sessionImplementor instanceof StatelessSession) {
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/id/ForeignGenerator.java]: could you squash the git commits into the main one?

**Comment by NathanQingyangXu** [comment]
Will updated accordingly pretty soon. Thanks for your review.

File: hibernate-core/src/main/java/org/hibernate/id/ForeignGenerator.java (hunk 2 above)
Comment (NathanQingyangXu) [reply hibernate-core/src/main/java/org/hibernate/id/ForeignGenerator.java]: I squashed and moved static import to the last. Still need a testing case.

Comment (NathanQingyangXu) [reply hibernate-core/src/main/java/org/hibernate/id/ForeignGenerator.java]: I created a testing case and squashed everything. Enjoy!

File: hibernate-core/src/test/java/org/hibernate/test/id/ForeignGeneratorTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+		inStatelessSession(statelessSession -> {
+
+			Transaction tx = statelessSession.beginTransaction();
+
+			Product product = new Product();
+
+			ProductDetails productDetails = new ProductDetails( product );
+
+			statelessSession.insert( productDetails );
```
Comment (NathanQingyangXu) [inline hibernate-core/src/test/java/org/hibernate/test/id/ForeignGeneratorTest.java]: I intentionally leave out `statelessSession.insert(product)` to cover an important logic branch for inserting reference entity automatically.

File: hibernate-core/src/main/java/org/hibernate/id/ForeignGenerator.java (hunk 2 above)
Comment (NathanQingyangXu) [reply hibernate-core/src/main/java/org/hibernate/id/ForeignGenerator.java]: @Sanne could you take a new look? I think it has been improved enormously after a well-thought testing case was added.

**Comment by sebersole** [comment]
Applied upstream.  Thanks again Nathan!

**Pull request closed without merging**